# Naive Bayes 
Prepare a classification model using Naive Bayes 
for salary data 

Data Description:

age -- age of a person
workclass	-- A work class is a grouping of work 
education	-- Education of an individuals	
maritalstatus -- Marital status of an individulas	
occupation	 -- occupation of an individuals
relationship -- 	
race --  Race of an Individual
sex --  Gender of an Individual
capitalgain --  profit received from the sale of an investment	
capitalloss	-- A decrease in the value of a capital asset
hoursperweek -- number of hours work per week	
native -- Native of an individual
Salary -- salary of an individual


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import os
import warnings
warnings.filterwarnings('ignore')

from pandas.plotting import scatter_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split 
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn import metrics
import statsmodels.api as sm

from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, plot_confusion_matrix


In [3]:
salarydata_train = pd.read_csv("SalaryData_Train.csv")
salarydata_test = pd.read_csv('SalaryData_Test.csv')


In [13]:
salarydata_train.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [14]:
salarydata_test.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [15]:
salarydata_train.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [16]:
salarydata_test.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [17]:
salarydata_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [18]:
salarydata_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


# Data Cleaning

In [44]:
#Finding the special characters in the data frame 
salarydata_test.isin(['?','\d']).sum(axis=0)

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [47]:
#Finding the special characters in the data frame 
salarydata_test.isin(['?']).sum(axis=0)

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [8]:
#find Categorical Variables
categorical=[var for var in salarydata_train.columns if salarydata_train[var].dtype=='O']
print('There are {} categorical variables\n'.format(len(categorical)))
print('The categorical variables are :\n\n', categorical)

There are 9 categorical variables

The categorical variables are :

 ['workclass', 'education', 'maritalstatus', 'occupation', 'relationship', 'race', 'sex', 'native', 'Salary']


In [13]:
# view the categorical variables

salarydata_train[categorical].head()

workclass   education        maritalstatus          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex          native  Salary  
0   Not-in-family   White     Male   United-States   <=50K  
1         Husband   White     Male   United-States   <=50K  
2   Not-in-family   White     Male   United-States   <=50K  
3         Husband   Black     Male   United-States   <=50K  
4            Wife   Black   Female            Cuba   <=50K

In [14]:
# check missing values in categorical variables
salarydata_train[categorical].isnull().sum()

workclass        0
education        0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
native           0
Salary           0
dtype: int64

In [15]:
# view frequency counts of values in categorical variables

for var in categorical: 
  
    print(salarydata_train[var].value_counts())

 Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64
 HS-grad         9840
 Some-college    6677
 Bachelors       5044
 Masters         1627
 Assoc-voc       1307
 11th            1048
 Assoc-acdm      1008
 10th             820
 7th-8th          557
 Prof-school      542
 9th              455
 12th             377
 Doctorate        375
 5th-6th          288
 1st-4th          151
 Preschool         45
Name: education, dtype: int64
 Married-civ-spouse       14065
 Never-married             9725
 Divorced                  4214
 Separated                  939
 Widowed                    827
 Married-spouse-absent      370
 Married-AF-spouse           21
Name: maritalstatus, dtype: int64
 Prof-specialty       4038
 Craft-repair         4030
 Exec-managerial      3992
 Adm-clerical         3721
 Sales                3584
 Other

In [16]:
# view frequency distribution of categorical variables

for var in categorical: 
    
    print(salarydata_train[var].value_counts()/np.float(len(salarydata_train)))
 

 Private             0.738868
 Self-emp-not-inc    0.082855
 Local-gov           0.068532
 State-gov           0.042406
 Self-emp-inc        0.035609
 Federal-gov         0.031266
 Without-pay         0.000464
Name: workclass, dtype: float64
 HS-grad         0.326249
 Some-college    0.221379
 Bachelors       0.167236
 Masters         0.053944
 Assoc-voc       0.043334
 11th            0.034747
 Assoc-acdm      0.033421
 10th            0.027187
 7th-8th         0.018468
 Prof-school     0.017970
 9th             0.015086
 12th            0.012500
 Doctorate       0.012433
 5th-6th         0.009549
 1st-4th         0.005006
 Preschool       0.001492
Name: education, dtype: float64
 Married-civ-spouse       0.466331
 Never-married            0.322436
 Divorced                 0.139717
 Separated                0.031133
 Widowed                  0.027420
 Married-spouse-absent    0.012267
 Married-AF-spouse        0.000696
Name: maritalstatus, dtype: float64
 Prof-specialty       0.13388

In [17]:
# check labels in workclass variable

salarydata_train.workclass.unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' Self-emp-inc', ' Without-pay'], dtype=object)

In [18]:
# check frequency distribution of values in workclass variable

salarydata_train.workclass.value_counts()

 Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64

In [19]:
# check for cardinality in categorical variables

for var in categorical:
    
    print(var, ' contains ', len(salarydata_train[var].unique()), ' labels')


workclass  contains  7  labels
education  contains  16  labels
maritalstatus  contains  7  labels
occupation  contains  14  labels
relationship  contains  6  labels
race  contains  5  labels
sex  contains  2  labels
native  contains  40  labels
Salary  contains  2  labels


Explore Numerical Variables

In [20]:
# find numerical variables

numerical = [var for var in salarydata_train.columns if salarydata_train[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 5 numerical variables

The numerical variables are : ['age', 'educationno', 'capitalgain', 'capitalloss', 'hoursperweek']


In [21]:
# view the numerical variables

salarydata_train[numerical].head()

age  educationno  capitalgain  capitalloss  hoursperweek
0   39           13         2174            0            40
1   50           13            0            0            13
2   38            9            0            0            40
3   53            7            0            0            40
4   28           13            0            0            40

In [24]:
# check missing values in numerical variables

salarydata_train[numerical].isnull().sum()

age             0
educationno     0
capitalgain     0
capitalloss     0
hoursperweek    0
dtype: int64

In [25]:
X = salarydata_train.drop(['Salary'], axis=1)

y = salarydata_train['Salary']

In [26]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)


In [27]:
# check the shape of X_train and X_test

X_train.shape, X_test.shape

((21112, 13), (9049, 13))

Feature Engineering

In [28]:
X_train.dtypes

age               int64
workclass        object
education        object
educationno       int64
maritalstatus    object
occupation       object
relationship     object
race             object
sex              object
capitalgain       int64
capitalloss       int64
hoursperweek      int64
native           object
dtype: object

In [29]:
X_test.dtypes

age               int64
workclass        object
education        object
educationno       int64
maritalstatus    object
occupation       object
relationship     object
race             object
sex              object
capitalgain       int64
capitalloss       int64
hoursperweek      int64
native           object
dtype: object

In [30]:
# display categorical variables

categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']

categorical

['workclass',
 'education',
 'maritalstatus',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native']

In [32]:
# display numerical variables

numerical = [col for col in X_train.columns if X_train[col].dtypes != 'O']

numerical

['age', 'educationno', 'capitalgain', 'capitalloss', 'hoursperweek']

In [36]:
# print percentage of missing values in the Numerical variables in training set

X_train[numerical].isnull().mean()

age             0.0
educationno     0.0
capitalgain     0.0
capitalloss     0.0
hoursperweek    0.0
dtype: float64

In [ ]:
# print percentage of missing values in the categorical variables in training set

X_train[categorical].isnull().mean()

In [35]:
# print categorical variables with missing data

for col in categorical:
    if X_train[col].isnull().mean()>0:
        print(col, (X_train[col].isnull().mean()))

In [37]:
# impute missing categorical variables with most frequent value

for df2 in [X_train, X_test]:
    df2['workclass'].fillna(X_train['workclass'].mode()[0], inplace=True)
    df2['occupation'].fillna(X_train['occupation'].mode()[0], inplace=True)
    df2['native'].fillna(X_train['native'].mode()[0], inplace=True)  


In [38]:
# check missing values in categorical variables in X_train

X_train[categorical].isnull().sum()

workclass        0
education        0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
native           0
dtype: int64

In [39]:
# check missing values in categorical variables in X_test

X_test[categorical].isnull().sum()

workclass        0
education        0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
native           0
dtype: int64

In [40]:
# check missing values in X_train

X_train.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
dtype: int64

In [41]:
# check missing values in X_test

X_test.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
dtype: int64

In [42]:
# print categorical variables

categorical

['workclass',
 'education',
 'maritalstatus',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native']

In [43]:
X_train[categorical].head()

workclass      education        maritalstatus  \
8166          Local-gov   Some-college   Married-civ-spouse   
7138            Private   Some-college        Never-married   
437             Private        HS-grad        Never-married   
5436            Private        HS-grad   Married-civ-spouse   
6541   Self-emp-not-inc        HS-grad   Married-civ-spouse   

             occupation    relationship    race    sex          native  
8166    Protective-serv         Husband   White   Male   United-States  
7138      Other-service       Own-child   White   Male   United-States  
437    Transport-moving   Not-in-family   White   Male   United-States  
5436       Craft-repair         Husband   White   Male   United-States  
6541       Tech-support         Husband   White   Male   United-States

In [48]:
!pip install category_encoders


In [49]:
# import category encoders

import category_encoders as ce

In [50]:
# encode remaining variables with one-hot encoding

encoder = ce.OneHotEncoder(cols=['workclass', 'education', 'maritalstatus', 'occupation', 'relationship', 
                                 'race', 'sex', 'native'])

X_train = encoder.fit_transform(X_train)

X_test = encoder.transform(X_test)

In [51]:
X_train

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
8166    54            1            0            0            0            0   
7138    21            0            1            0            0            0   
437     30            0            1            0            0            0   
5436    42            0            1            0            0            0   
6541    37            0            0            1            0            0   
...    ...          ...          ...          ...          ...          ...   
13123   66            1            0            0            0            0   
19648   26            0            1            0            0            0   
9845    50            0            1            0            0            0   
10799   49            0            0            1            0            0   
2732    34            0            1            0            0            0   

       workclass_6  workclass_7  education_1  education_2  ...  native_31  \
8166             0            0            1            0  ...          0   
7138             0            0            1            0  ...          0   
437              0            0            0            1  ...          0   
5436             0            0            0            1  ...          0   
6541             0            0            0            1  ...          0   
...            ...          ...          ...          ...  ...        ...   
13123            0            0            0            0  ...          0   
19648            0            0            0            1  ...          0   
9845             0            0            0            0  ...          0   
10799            0            0            0            0  ...          0   
2732             0            0            0            1  ...          0   

       native_32  native_33  native_34  native_35  native_36  native_37  \
8166           0          0          0          0          0          0   
7138           0          0          0          0          0          0   
437            0          0          0          0          0          0   
5436           0          0          0          0          0          0   
6541           0          0          0          0          0          0   
...          ...        ...        ...        ...        ...        ...   
13123          0          0          0          0          0          0   
19648          0          0          0          0          0          0   
9845           0          0          0          0          0          0   
10799          0          0          0          0          0          0   
2732           0          0          0          0          0          0   

       native_38  native_39  native_40  
8166           0          0          0  
7138           0          0          0  
437            0          0          0  
5436           0          0          0  
6541           0          0          0  
...          ...        ...        ...  
13123          0          0          0  
19648          0          0          0  
9845           0          0          0  
10799          0          0          0  
2732           0          0          0  

[21112 rows x 102 columns]

In [52]:
X_test

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
25338   21            0            1            0            0            0   
18840   21            0            1            0            0            0   
8391    56            0            1            0            0            0   
18258   43            1            0            0            0            0   
16669   53            0            0            0            1            0   
...    ...          ...          ...          ...          ...          ...   
28655   24            0            0            1            0            0   
14796   18            0            1            0            0            0   
20728   49            0            1            0            0            0   
19297   37            0            1            0            0            0   
8577    23            0            1            0            0            0   

       workclass_6  workclass_7  education_1  education_2  ...  native_31  \
25338            0            0            0            1  ...          0   
18840            0            0            0            0  ...          0   
8391             0            0            0            0  ...          0   
18258            0            0            1            0  ...          0   
16669            0            0            0            0  ...          0   
...            ...          ...          ...          ...  ...        ...   
28655            0            0            1            0  ...          0   
14796            0            0            1            0  ...          0   
20728            0            0            0            0  ...          0   
19297            0            0            0            1  ...          0   
8577             0            0            1            0  ...          0   

       native_32  native_33  native_34  native_35  native_36  native_37  \
25338          0          0          0          0          0          0   
18840          0          0          0          0          0          0   
8391           0          0          0          0          0          0   
18258          0          0          0          0          0          0   
16669          1          0          0          0          0          0   
...          ...        ...        ...        ...        ...        ...   
28655          0          0          0          0          0          0   
14796          0          0          0          0          0          0   
20728          0          0          0          0          0          0   
19297          0          0          0          0          0          0   
8577           0          0          0          0          0          0   

       native_38  native_39  native_40  
25338          0          0          0  
18840          0          0          0  
8391           0          0          0  
18258          0          0          0  
16669          0          0          0  
...          ...        ...        ...  
28655          0          0          0  
14796          0          0          0  
20728          0          0          0  
19297          0          0          0  
8577           0          0          0  

[9049 rows x 102 columns]

We now have training and testing set ready for model building. Before that, we should map all the feature variables onto the same scale. It is called feature scaling.

In [53]:
cols = X_train.columns

In [56]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)

In [57]:
X_train

array([[ 0.89473684,  1.        , -1.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.84210526,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.36842105,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [ 0.68421053,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.63157895,  0.        , -1.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.15789474,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ]])

In [58]:
X_train = pd.DataFrame(X_train, columns=[cols])
X_train

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0      0.894737         1.0        -1.0         0.0         0.0         0.0   
1     -0.842105         0.0         0.0         0.0         0.0         0.0   
2     -0.368421         0.0         0.0         0.0         0.0         0.0   
3      0.263158         0.0         0.0         0.0         0.0         0.0   
4      0.000000         0.0        -1.0         1.0         0.0         0.0   
...         ...         ...         ...         ...         ...         ...   
21107  1.526316         1.0        -1.0         0.0         0.0         0.0   
21108 -0.578947         0.0         0.0         0.0         0.0         0.0   
21109  0.684211         0.0         0.0         0.0         0.0         0.0   
21110  0.631579         0.0        -1.0         1.0         0.0         0.0   
21111 -0.157895         0.0         0.0         0.0         0.0         0.0   

      workclass_6 workclass_7 education_1 education_2  ... native_31  \
0             0.0         0.0         1.0         0.0  ...       0.0   
1             0.0         0.0         1.0         0.0  ...       0.0   
2             0.0         0.0         0.0         1.0  ...       0.0   
3             0.0         0.0         0.0         1.0  ...       0.0   
4             0.0         0.0         0.0         1.0  ...       0.0   
...           ...         ...         ...         ...  ...       ...   
21107         0.0         0.0         0.0         0.0  ...       0.0   
21108         0.0         0.0         0.0         1.0  ...       0.0   
21109         0.0         0.0         0.0         0.0  ...       0.0   
21110         0.0         0.0         0.0         0.0  ...       0.0   
21111         0.0         0.0         0.0         1.0  ...       0.0   

      native_32 native_33 native_34 native_35 native_36 native_37 native_38  \
0           0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1           0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2           0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3           0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4           0.0       0.0       0.0       0.0       0.0       0.0       0.0   
...         ...       ...       ...       ...       ...       ...       ...   
21107       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
21108       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
21109       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
21110       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
21111       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

      native_39 native_40  
0           0.0       0.0  
1           0.0       0.0  
2           0.0       0.0  
3           0.0       0.0  
4           0.0       0.0  
...         ...       ...  
21107       0.0       0.0  
21108       0.0       0.0  
21109       0.0       0.0  
21110       0.0       0.0  
21111       0.0       0.0  

[21112 rows x 102 columns]

In [59]:
X_test = pd.DataFrame(X_test, columns=[cols])
X_test

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0    -0.842105         0.0         0.0         0.0         0.0         0.0   
1    -0.842105         0.0         0.0         0.0         0.0         0.0   
2     1.000000         0.0         0.0         0.0         0.0         0.0   
3     0.315789         1.0        -1.0         0.0         0.0         0.0   
4     0.842105         0.0        -1.0         0.0         1.0         0.0   
...        ...         ...         ...         ...         ...         ...   
9044 -0.684211         0.0        -1.0         1.0         0.0         0.0   
9045 -1.000000         0.0         0.0         0.0         0.0         0.0   
9046  0.631579         0.0         0.0         0.0         0.0         0.0   
9047  0.000000         0.0         0.0         0.0         0.0         0.0   
9048 -0.736842         0.0         0.0         0.0         0.0         0.0   

     workclass_6 workclass_7 education_1 education_2  ... native_31 native_32  \
0            0.0         0.0         0.0         1.0  ...       0.0       0.0   
1            0.0         0.0         0.0         0.0  ...       0.0       0.0   
2            0.0         0.0         0.0         0.0  ...       0.0       0.0   
3            0.0         0.0         1.0         0.0  ...       0.0       0.0   
4            0.0         0.0         0.0         0.0  ...       0.0       1.0   
...          ...         ...         ...         ...  ...       ...       ...   
9044         0.0         0.0         1.0         0.0  ...       0.0       0.0   
9045         0.0         0.0         1.0         0.0  ...       0.0       0.0   
9046         0.0         0.0         0.0         0.0  ...       0.0       0.0   
9047         0.0         0.0         0.0         1.0  ...       0.0       0.0   
9048         0.0         0.0         1.0         0.0  ...       0.0       0.0   

     native_33 native_34 native_35 native_36 native_37 native_38 native_39  \
0          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
...        ...       ...       ...       ...       ...       ...       ...   
9044       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
9045       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
9046       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
9047       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
9048       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

     native_40  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  
...        ...  
9044       0.0  
9045       0.0  
9046       0.0  
9047       0.0  
9048       0.0  

[9049 rows x 102 columns]

In [60]:
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB


# instantiate the model
gnb = GaussianNB()


# fit the model
gnb.fit(X_train, y_train)

GaussianNB()

In [61]:
y_pred = gnb.predict(X_test)

y_pred

array([' <=50K', ' <=50K', ' <=50K', ..., ' <=50K', ' <=50K', ' >50K'],
      dtype='<U6')

Check accuracy score

In [62]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.7995


Here, y_test are the true class labels and y_pred are the predicted class labels in the test-set.

Compare the train-set and test-set accuracy

In [63]:
y_pred_train = gnb.predict(X_train)

y_pred_train

array([' >50K', ' <=50K', ' <=50K', ..., ' <=50K', ' >50K', ' <=50K'],
      dtype='<U6')

In [64]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 0.8023


Check for overfitting and underfitting

In [65]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(gnb.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(gnb.score(X_test, y_test)))

Training set score: 0.8023
Test set score: 0.7995


Compare model accuracy with null accuracy

In [66]:
# check class distribution in test set

y_test.value_counts()

 <=50K    6798
 >50K     2251
Name: Salary, dtype: int64

In [67]:
# check null accuracy score

null_accuracy = (7407/(7407+2362))

print('Null accuracy score: {0:0.4f}'. format(null_accuracy))

Null accuracy score: 0.7582


We can see that our model accuracy score is 0.8023 but null accuracy score is 0.7582. So, we can conclude that our Gaussian Naive Bayes Classification model is doing a very good job in predicting the class labels.